### PR TITLE
NMS-13802: Add docs for Device Config

### PR DESCRIPTION
**Pull request by cgorantla**
Thanks for taking time to contribute!

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-13802

### CODE REVIEW

File: docs/modules/operation/pages/device-config-backup/configuration.adoc [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+| uei.opennms.org/deviceconfig/configRetrievalSucceeded
+| Config retrieval succeeded on a given service during the last poll on a given interface.
+
+|===
+
+== Retrieve Device Config without Monitor
+
+Retrieving Device Config can also be triggered outside Monitor schedule with Rest API.
+see xref:development:rest/device_config.adoc[Device Config Monitor].
```
Comment (Bonrob2) [inline docs/modules/operation/pages/device-config-backup/configuration.adoc:77]:
```suggestion
(See xref:development:rest/device_config.adoc[Device Config Monitor].)
```

File: docs/modules/operation/pages/device-config-backup/configuration.adoc [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+
+== Retrieve/Trigger Device Config Backup from Karaf shell
```
Comment (Bonrob2) [inline docs/modules/operation/pages/device-config-backup/configuration.adoc:80]:
```suggestion
== Retrieve/trigger device configuration backup from Karaf shell
```

File: docs/modules/operation/pages/device-config-backup/configuration.adoc [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+Device Config can also be triggered/retrieved from karaf shell.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/device-config-backup/configuration.adoc:82]:
```suggestion
You can also trigger/retrieve a device configuration from the Karaf shell.
```

File: docs/modules/reference/nav.adoc [hunk 4]
```diff
@@ -35,6 +35,7 @@
 ** xref:service-assurance/monitors/CiscoIpSlaMonitor.adoc[]
 ** xref:service-assurance/monitors/CiscoPingMibMonitor.adoc[]
 ** xref:service-assurance/monitors/CitrixMonitor.adoc[]
+** xref:service-assurance/monitors/DeviceConfiMonitor.adoc[]
```
Comment (Bonrob2) [inline docs/modules/reference/nav.adoc:38]: verifying spelling ... the file is DeviceConfiMonitor.adoc, but chapter title is "DeviceConfigMonitor" (see comment line 2 in DeviceConfiMonitor.adoc file.

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc [hunk 5]
```diff
@@ -0,0 +1,57 @@
+
+= DeviceConfigMonitor
```
Comment (Bonrob2) [inline docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc:2]:
```suggestion
= DeviceConfigMonitor
```
Is this correct name? Could we rename .adoc file and the nav file entry  to include the "g"?

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+This Monitor tries to retrieve config from the network device and will show corresponding status based on whether config retrieval succeeded or not.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc:4]:
```suggestion
This monitor tries to retrieve configuration from the network device and shows the status of that retrieval (success, failed, and so on).
```

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+|===
+
+| Parameter
+| Description
+| Default value
+
+3+| *Required*
+
+| script-file
+| Script file that needs to be used to retrieve config.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc:27]:
```suggestion
| Script file to use to retrieve configuration.
```

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+| `-`
+
+3+| *Optional*
+
+| config-type
+| If Device has multiple configs, config-type should be specified with each service.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc:33]:
```suggestion
| If device has multiple configurations, specify the config-type with each service.
```

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+| default
+
+| schedule
+| Schedule for fetching config in the form of cron expression.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc:37]:
```suggestion
| Schedule to fetch configuration, in the form of cron expression.
```

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+| 0 0 0 * * ?
+
+|===
+
+
+== Examples
+
+example configuration on how to configure the DeviceConfigMonitor in the `poller-configuration.xml`.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc:45]:
```suggestion
Example for how to configure the DeviceConfigMonitor in the `poller-configuration.xml`.
```

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+[source, xml]
+----
+<parameter key="schedule" value="${requisition:dcb:schedule|0 0 0 * * ?}"/>
+<parameter key="script-file" value="${requisition:script-file|juniper-vsrx-default}.dcb"/>
+----
+
+== Schedule and interval
+
+For Device Config Monitor service, interval is configured to be 300secs and schedule is a cron expression. Default
```
Comment (Bonrob2) [inline docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc:55]:
```suggestion
The Device Config Monitor service uses an the interval of 300secs and a cron expression for the schedule.
```

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+cron expression runs every 24 hrs at 12.00am (0 0 0 * * ?). Monitor retrieves the Device config when the next trigger
```
Comment (Bonrob2) [inline docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc:56]:
```suggestion
The default cron expression runs every 24 hrs at 12.00am (0 0 0 * * ?). 
```

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+from schedule falls within the interval.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/service-assurance/monitors/DeviceConfiMonitor.adoc:57]:
```suggestion
The monitor retrieves the device configuration when the next trigger from the schedule falls within the interval.
```

File: docs/modules/development/pages/rest/device_config.adoc [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+[options="header, %autowidth"]
+|===
+| Resource
+| Method
+| Description
+
+| `/device-config`
+| GET
+| Retrieves all the configs sorted by lastUpdated.
```
Comment (Bonrob2) [inline docs/modules/development/pages/rest/device_config.adoc:16]:
```suggestion
| Retrieves all the configurations, sorted by lastUpdated.
```

File: docs/modules/development/pages/rest/device_config.adoc [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+| This API also supports following query params: limit, offset, orderBy, order, deviceName, ipAddress, configType,
+| createdAfter, createdBefore.
+
+| `/device-config/\{id}`
+| GET
+| Retrieves the config for the specific id (Database ID).
+
+| `/device-config/interface/\{id}`
+| GET
+| Retrieves all the configs for the specific interface.
```
Comment (Bonrob2) [inline docs/modules/development/pages/rest/device_config.adoc:26]:
```suggestion
| Retrieves all the configurations for the specific interface.
```

File: docs/modules/development/pages/rest/device_config.adoc [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+| `/device-config/latest`
+| GET
+| Retrieves the latest config for all the devices.
```
Comment (Bonrob2) [inline docs/modules/development/pages/rest/device_config.adoc:30]:
```suggestion
| Retrieves the latest configuration for all the devices.
```

File: docs/modules/development/pages/rest/device_config.adoc [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+| `/device-config/backup`
+| POST
+| Trigger backup for a given interface and location.
+
+|===
+
+== Examples:
+
+.Trigger Device Config backup
```
Comment (Bonrob2) [inline docs/modules/development/pages/rest/device_config.adoc:44]:
```suggestion
.Trigger device configuration backup
```

**Review by Bonrob2 — CHANGES_REQUESTED** [review]
Looking good. We may need to restructure the content once the UI documentation side is done, but we can do that as part of that Jira.

File: docs/modules/development/pages/rest/device_config.adoc [hunk 18]
```diff
@@ -0,0 +1,49 @@
+[[ga-development-rest-device-config]]
+= Device Config Rest API
+
+Use the Device Config Rest API to fetch existing configs and trigger retrieval of the device configs and is in `api/v2` endpoint.
```
[the remark this answers is not in the document]
Comment (cgorantla) [reply docs/modules/development/pages/rest/device_config.adoc]: Removed

File: docs/modules/operation/pages/device-config-backup/configuration.adoc [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+----
+
+.Configure Device Config Sink feature via Karaf shell on Minion
+[source, console]
+----
+ssh -p 8201 admin@localhost
+----
+
+
+.Configure opennms-deviceconfig-sink feature
```
[the remark this answers is not in the document]
Comment (cgorantla) [reply docs/modules/operation/pages/device-config-backup/configuration.adoc]: It's applicable for both. Added `({page-component-title}/Minion)`

File: docs/modules/operation/pages/device-config-backup/configuration.adoc [hunk 20]
```diff
@@ -1,35 +1,35 @@
 == Poller configuration
```
Comment (Bonrob2) [inline docs/modules/operation/pages/device-config-backup/configuration.adoc:1]: = Poller configuration

Title level needs to be updated, as the table of contents did not build properly.

File: docs/modules/operation/pages/device-config-backup/configuration.adoc [hunk 21]
```diff
@@ -71,15 +71,15 @@ during the last poll on a given interface
 
 |===
 
-== Retrieve Device Config without Monitor
+== Retrieve device configuration without monitor
 
-Retrieving Device Config can also be triggered outside Monitor schedule with Rest API.
-see xref:development:rest/device_config.adoc[Device Config Monitor].
+You can use the Rest API to trigger device configuration retrieval outside of a monitor schedule.
+(See xref:development:rest/device_config.adoc[Device Config Monitor].)
```
Comment (Bonrob2) [inline docs/modules/operation/pages/device-config-backup/configuration.adoc:77]: (See xref:development:rest/device_config.adoc[Device Config Rest API].

File: docs/modules/reference/pages/service-assurance/monitors/DeviceConfigMonitor.adoc [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+
+[source, xml]
+----
+<parameter key="schedule" value="${requisition:dcb:schedule|0 0 0 * * ?}"/>
+<parameter key="script-file" value="${requisition:script-file|juniper-vsrx-default}.dcb"/>
+----
+
+== Schedule and interval
+
+The Device Config Monitor service uses an the interval of 300secs and a cron expression for the schedule.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/service-assurance/monitors/DeviceConfigMonitor.adoc:55]: The Device Config Monitor service uses an interval of 300secs and a cron expression for the schedule.

File: docs/modules/operation/pages/device-config-backup/configuration.adoc [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+
+| uei.opennms.org/deviceconfig/configRetrievalSucceeded
+| Config retrieval succeeded on a given service during the last poll on a given interface.
+
+|===
+
+== Retrieve device configuration without monitor
+
+You can use the Rest API to trigger device configuration retrieval outside of a monitor schedule.
+(See xref:development:rest/device_config.adoc[Device Config Rest API].
```
Comment (Bonrob2) [inline docs/modules/operation/pages/device-config-backup/configuration.adoc:77]:
```suggestion
(See xref:development:rest/device_config.adoc[Device Config Rest API].)
```

**Review by Bonrob2 — APPROVED** [review]
I missed the closing parenthesis, so just added it and committed it. Everything else looks good. Thanks!

**Pull request merged**